File: src/main/java/com/baidu/algorithm/dp/_132_Palindrome_Partitioning_II.py
class Solution(object):
    def minCut(self, s):
        """
        :type s: str
        :rtype: int
        """
        if not s or len(s) < 2:
            return 0

        n = len(s)
        dp = [[False] * n for _ in xrange(n)]
        cut = [0] * n

        for i in xrange(n):
            min_val = i
            for j in xrange(i + 1):
                if s[i] == s[j] and (j + 1 > i - 1 or dp[j + 1][i - 1]):
                    dp[j][i] = True
                    min_val = 0 if j == 0 else min(min_val, cut[j - 1] + 1)
            cut[i] = min_val

        return cut[n - 1]


if __name__ == '__main__':
    s = Solution()
    s.minCut("abbab")

File: src/main/java/com/baidu/algorithm/dp/_132_Palindrome_Partitioning_II.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.dp;

/**
 * _132_Palindrome_Partitioning_II
 *
 * @author xuhaoran01
 */
public class _132_Palindrome_Partitioning_II {
    public int minCut(String s) {
        if(s == null || s.isEmpty()) {
            return 0;
        }

        int n = s.length();
        boolean[][] dp = new boolean[n][n];
        for(int i = 0; i < n; i++) {
            for(int j = 0; j <= i; j++) {
                if(s.charAt(i) == s.charAt(j) && (j + 1 > i - 1 || dp[j + 1][i - 1])) {
                    dp[j][i] = true;
                }
            }
        }

        int[] cut = new int[n];
        for(int i = 1; i < n; i++) {
            cut[i] = i + 1;
            for(int j = i; j >= 0; j--) {
                if(dp[j][i]) {
                    cut[i] = Math.min(j == 0 ? 0 : cut[j - 1] + 1, cut[i]);
                }
            }
        }

        return cut[n - 1];
    }

    public static void main(String[] args) {
        new _132_Palindrome_Partitioning_II().minCut("aab");
    }
}
